### ISSUE TITLE
missing_docs does not trigger for types only used in exported fields of enumerations

### ISSUE THREAD

**Issue by mathstuf** [rust-lang/rust#39162]
IMO, this should be a compilation failure because although `Foo` is not explicitly exported, it is publicly visible via the `Bar::A` variant:

`lib.rs`:

```rust
#[deny(missing_docs)]

mod types;

pub use types::Bar;
```

`types.rs`:

```rust
pub struct Foo;

/// A bar.
pub enum Bar {
    /// A `Bar` variant.
    A(Foo),
}
```

Oddly enough, it seems to also be happy with the missing crate-level docs…

Using 1.14.0 (from Fedora's repository).

**Comment by mathstuf** [comment]
Hmm, even documenting it does not make it hyperlinked in the generated docs. What is the expected behavior here?

**Comment by durka** [comment]
You need to make the `types` module public.

**Comment by mathstuf** [comment]
Yeah, that makes the errors appear (as does re-exporting `Foo` from `lib.rs`). I'm not in front of a compiler right now, but IIRC, removing the `pub` from `Foo` is a compiler error; why is it not also a doc error?

**Comment by mathstuf** [comment]
Yep, removing the `pub` from `Foo` makes it error out that it can't leak private types. So, should the same behavior apply for `missing_docs` (since this is "leaking an undocumented public type")?

**Comment by Mark-Simulacrum** [comment]
Updated description to include `!` in #![deny(missing_docs)]`, which made the lint error on the crate documentation, but still no errors about `Foo` itself.

**Comment by spacejam** [comment]
It was a bit of a surprise when I realized an important type that users are exposed to in one of my libraries was not properly documented, despite the crate specifying `#![deny(missing_docs)]`. My use case looked a bit like [this](https://play.rust-lang.org/?gist=20ed101984d17e4209eb56932075395c&version=stable).
